File: Baekjoon/java/17141.java
import java.util.*;
import java.io.*;

class Pair {
	int x;
	int y;
	public Pair(int x, int y) {
		this.x = x;
		this.y = y;
	}
}

public class Main {
	
	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	StringTokenizer st;
	StringBuilder sb = new StringBuilder();
	int n, m;
	int[][] arr;
	List<Pair> vp = new ArrayList<>();
	Stack<Integer> stack  = new Stack<>();
	boolean[] check;
	Queue<Pair> q = new LinkedList<>();
	int[][] vis;
	final int[] dx = {0,0,1,-1};
	final int[] dy = {1,-1,0,0};
	int ans = Integer.MAX_VALUE;
	
	void init() {
		try {
			st = new StringTokenizer(br.readLine(), " ");
			n = Integer.parseInt(st.nextToken());
			m = Integer.parseInt(st.nextToken());
			arr = new int[n][n];
			vis = new int[n][n];
			for(int i=0; i<n; i++) {
				st = new StringTokenizer(br.readLine(), " ");
				for(int j=0; j<n; j++) {
					arr[i][j] = Integer.parseInt(st.nextToken());
					if(arr[i][j]==2) {
						vp.add(new Pair(i, j));
						arr[i][j] = 0;
					}
				}
			}
			check = new boolean[vp.size()];
			
		} catch(IOException e) {}
	}
	
	void time() {
		int max = vis[0][0];
		boolean flag = false;
		for(int i=0; i<n; i++) {
			for(int j=0; j<n; j++) {
				if(vis[i][j]==-1 && arr[i][j]==0) {
					flag = true;
				}
				max = max<vis[i][j]?vis[i][j]:max;
			}
		}
		if(!flag) {
			ans = max<ans?max:ans;
		}
//		System.out.println("ans="+ans);
	}
	
	void bfs() {		
		while(!q.isEmpty()) {
			Pair p = q.poll();
			int x = p.x;
			int y = p.y;
			
//			if(vis[x][y]==-1) 
//				vis[x][y] = 0;
			
			for(int k=0; k<4; k++) {
				int nx = x + dx[k];
				int ny = y + dy[k];				
				if(nx>=0 && nx<n && ny>=0 && ny<n) {
					if(vis[nx][ny]==-1 && arr[nx][ny]==0) {
						vis[nx][ny] = vis[x][y] + 1;
						q.add(new Pair(nx, ny));
					}
				}
			}
		}
	}	
	
	void dfs(int x) {
		if(stack.size()==m) {
			for(int i=0; i<n; i++)
				Arrays.fill(vis[i], -1);
//			System.out.print("virus : ");
			for(int item: stack) {
//				System.out.print("("+vp.get(item).x+", "+vp.get(item).y+")");
				q.add(new Pair(vp.get(item).x, vp.get(item).y));
				vis[vp.get(item).x][vp.get(item).y] = 0;
			}
//			System.out.println();
			bfs();
			System.out.println("===vis===");
			print(vis);
			time();
		}
		
		for(int i=x; i<vp.size(); i++) {
			if(!check[i]) {
				check[i] = true;
				stack.add(i);
				dfs(i+1);
				stack.pop();
				check[i] = false;
			}
		}
	}
	
	void solve() {
		init();
		dfs(0);
		System.out.println(ans==Integer.MAX_VALUE?-1:ans);
	}
	
	void print(int[][] arr) {
		for(int i=0; i<n; i++) {
			for(int j=0; j<n; j++) {
				if(arr[i][j]==-1) System.out.print("- ");
				else System.out.print(arr[i][j]+" ");
			}
			System.out.println();
		}
	}

	public static void main(String[] args) {
		new Main().solve();
	}
}

File: Baekjoon/cpp/17141.cpp
#include <iostream>
#include <algorithm>
#include <queue>
#include <vector>
using namespace std;

void init() {
	ios::sync_with_stdio(0);
	cin.tie(0); cout.tie(0);
}

const int MAX = 50 + 2;
int n, m, p_idx, ans = 987654321;
int arr[MAX][MAX], vis[MAX][MAX];
pair<int, int> place[10];
vector<int> v;
bool check[10];
queue<pair<int, int>> q;
int dx[4] = { 0,0,1,-1 };
int dy[4] = { 1,-1,0,0 };

void init_vis() {
	for (int i = 0; i < n; i++)
		for (int j = 0; j < n; j++)
			vis[i][j] = -1;
}

void bfs() {
	while (!q.empty()) {
		int x = q.front().first;
		int y = q.front().second;
		q.pop();

		for (int k = 0; k < 4; k++) {
			int nx = x + dx[k];
			int ny = y + dy[k];

			if (nx >= 0 && nx < n && ny >= 0 && ny < n) {
				if (vis[nx][ny] == -1 && (arr[nx][ny] == 0 || arr[nx][ny] == 2)) {
					vis[nx][ny] = vis[x][y] + 1;
					q.push({ nx, ny });
				}
			}
		}
	}
}

int get_time() {
	int min_time = -1;
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			if (vis[i][j] == -1 && arr[i][j] == 0) return -1;
			if (min_time < vis[i][j]) 
				min_time = vis[i][j];
		}
	}
	return min_time;
}

void dfs(int x) {	//place에 들어있는 장소중에 m개의 장소 고르기
	if (v.size() == m) {
		init_vis();

		for (int i = 0; i < v.size(); i++) {
			int xx = place[v[i]].first;
			int yy = place[v[i]].second;
			vis[xx][yy] = 0;
			q.push({ xx, yy });
		}

		bfs();

		int min_time = get_time();
		if (min_time != -1)
			ans = min(ans, min_time);
		
		return;
	}

	for (int i = x; i < p_idx; i++) {
		if (!check[i]) {
			check[i] = true;
			v.push_back(i);
			dfs(i + 1);
			check[i] = false;
			v.pop_back();
		}
	}
}

int main() {
	init();
	cin >> n >> m;
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			cin >> arr[i][j];
			if (arr[i][j] == 2) {
				place[p_idx].first = i;
				place[p_idx].second = j;
				p_idx++;
			}
		}
	}

	dfs(0);

	if (ans == 987654321) cout << -1;
	else cout << ans;
	
	return 0;
}
